osf/utils/users.py

```diff
@@ -0,0 +1,36 @@
+from django.utils import timezone
+from osf.models import OSFUser
+from django.core.exceptions import ValidationError
+from website.settings import DOMAIN
+
+def get_or_refresh_confirmation_link(uid):
+    u = OSFUser.load(uid)
+    if u.is_confirmed:
+        return {'error': f'User {uid} is already confirmed'}
+
+    if u.deleted or u.is_merged:
+        return {'error': f'User {uid} is deleted or merged'}
+
+    confirmation_link = None
+    if u.emails.filter(address=u.username).exists():
+        for token, data in u.email_verifications.items():
+            if data['email'] == u.username and data['expiration'] < timezone.now():
+                u.email_verifications[token]['expiration'] = timezone.now() + timezone.timedelta(days=30)
+                u.save()
+                confirmation_link = f'{DOMAIN.rstrip("/")}/confirm/{u._id}/{token}'
```

Comment:
Manually constructing the link here is a little fragile -- if we ever change the structure, this is likely be overlooked. It would be preferable to rely on the existing model function to `get_confirmation_url`, but... [continued in other comment]